### DISCUSSION TITLE
Is it possible to parallelize a loop?

### DISCUSSION THREAD

**Discussion by adalseno** [modular/modular#330] [Q&A] [answered]
Hi, I wrote a simple function to fill an array:
```
fn populate_mojo(b:Vect):
    for i in range(b.rows):
        b[i] = i
```
where Vect is a struct:
```
struct Vect:
    var data: DTypePointer[DType.ui64]
    var rows: Int

    fn __init__(inout self, rows: Int):
        self.data = DTypePointer[DType.ui64].alloc(rows)
        self.rows = rows

    fn __del__(owned self):
        self.data.free()
    
    @always_inline    
    fn len(self)->UI64:
        return self.rows

    fn zero(inout self):
        memset_zero(self.data, self.rows)

    @always_inline
    fn __getitem__(self, x: Int) -> UI64:
        return self.data.load(x)

    

    @always_inline
    fn __setitem__(self,  x: Int, val: UI64):
        return self.data.store( x, val)

```
called after creating the Vect:
```
let b = Vect(size)
populate_mojo(b)
```
I hope it makes sense (any hint is more than welcome).  The switch is not so trivial for a pure Python programmer, but the speed gain is impressive. It's much faster than its numpy counterpart (around 30 times faster for 1 million array on the playground Notebook):
```
def populate_numpy(size:int)->np.ndarray:
    b = np.empty((size),dtype=np.int64)
    for i in range(size):
        b[i] = i
    return b
```
I would like to know if it is possible to parallelize the loop and if it would improve the performance (in this case each loop operation is independent of the previous one, so they should work in parallel). Thanks in advance.

**Comment by czheo** [comment, marked as answer]
You can try https://docs.modular.com/mojo/MojoStdlib/Functional.html#parallelize

Something like this should work
```py
from Functional import parallelize

fn populate_mojo(b:Vect):
    @parameter
    fn process_row(i:Int):
        b[i] = i
    parallelize[process_row](b.rows)
```

**Reply by adalseno** [reply]
I will try, thanks. In the meanwhile, I implemented the same trivial function in Julia 1.7.2, and it's faster (more than twice as much; 12 ms vs. 27 ms for a 10 million element vector. Using Numba, instead, I get the same speed as Mojo):
```
function populate_array(size::Int)::AbstractVector{Int64}
    b = Vector{Int64}(undef,size)
    Threads.@threads for i=1:size
        b[i] = i
    end
    return b 
end
```
I just tried it, and it worked like a charm, with impressive speed. Now Mojo version is about twice as fast as Julia's (7 ms vs. 12 ms for a 10 million vector; 7 ms on the playground. I guess on my computer, it might be even faster). Amazing.
I had to slightly modify the struct since Mojo complained about the types:
```
struct Vect:
    var data: DTypePointer[DType.uint64]
    var rows: Int

    fn __init__(inout self, rows: Int):
        self.data = DTypePointer[DType.uint64].alloc(rows)
        self.rows = rows

    fn __del__(owned self):
        self.data.free()
    
    @always_inline    
    fn len(self)->UInt64:
        return self.rows

    fn zero(inout self):
        memset_zero(self.data, self.rows)

    @always_inline
    fn __getitem__(self, x: Int) -> UInt64:
        return self.data.load(x)

    

    @always_inline
    fn __setitem__(self,  x: Int, val: UInt64):
        return self.data.store( x, val)
```
